sqlmesh/core/plan/builder.py

```diff
@@ -673,6 +719,95 @@ def _ensure_new_env_with_changes(self) -> None:
                 "No changes were detected. Make a change or run with --include-unmodified to create a new environment without changes."
             )
 
+    def _upstream_directly_modified(
+        self, s_id: SnapshotId, directly_modified: t.Set[SnapshotId], dag: DAG[SnapshotId]
+    ) -> t.List[SnapshotId]:
+        return [
+            upstream_id for upstream_id in dag.upstream(s_id) if upstream_id in directly_modified
+        ]
+
+    def _validate_destructive_change(
+        self,
+        s_id: SnapshotId,
+        upstream_snapshot_ids: t.List[SnapshotId] = [],
+        force_check: bool = False,
+    ) -> None:
+        """Determine if a schema change requires a destructive operation"""
+        s_id_snapshot = t.cast(Snapshot, self._context_diff.snapshots.get(s_id))
+
+        if (
+            s_id_snapshot.is_model
+            and s_id_snapshot.name not in self._allow_destructive_models
+            and not s_id_snapshot.model.on_destructive_change.is_ignore
+        ):
+            warning_msg = f"PLAN TIME CHECK: Plan results in a destructive change to forward-only model '{s_id_snapshot.name}'s schema."
+            error_msg = f"{warning_msg} To allow this, change the model's `on_destructive_change` setting to `warn` or `ignore` or include it in the plan's `--allow-destructive-model` option."
+
+            def _raise_or_warn() -> None:
+                if s_id_snapshot.model.on_destructive_change.is_error:
+                    raise PlanError(error_msg)
+                logger.warning(warning_msg)
+
+            subdag = set([s_id, *upstream_snapshot_ids])
+            subdag_to_process = subdag.copy()
+            subdag_no_cols_to_types = set()
+
+            # get already classified snapshots
+            for id in subdag:
+                if id.name in self._snapshot_change_is_destructive:
+                    subdag_to_process.remove(id)
+                    value = self._snapshot_change_is_destructive[id.name]
+                    if value:
+                        _raise_or_warn()
```

Comment:
I'm a little confused. Don't we call `_raise_or_warn` when we set the key to the `_snapshot_change_is_destructive` on line 802? How do we end in this block? Do we warn more than once?

Reply:
Never mind, I think I get it. `_raise_or_warn` always checks the target snapshot and not the one for which the `_snapshot_change_is_destructive` was set.

Reply:
Right - the approach is "as soon as we determine that any of the target snapshot or an upstream has a destructive change, _raise_or_warn() for the target snapshot instead of looking at more snapshots"